File: back-end/c#/TaskManagement/BLL/WorkerHoursService.cs
using BOL;
using DAL;
using MySql.Data.MySqlClient;
using System;
using System.Collections.Generic;

namespace BLL
{
    public static class WorkerHoursService
    {
        public static List<WorkerHours> GetAllWorkerHours(int workerId)
        {
            string query;
            query = $"{GetWorkerHoursQuery()} WHERE worker_id ={workerId} AND w.is_complete=0;";
            List<WorkerHours> workerHoursList = GetWorkersHours(query);
            return workerHoursList;
        }

        public static bool EditWorkerHours(WorkerHours workerHours)
        {
            try
            {
                decimal presenceHoursSum = Convert.ToDecimal(PresenceHoursService.GetPresenceHoursSum(workerHours.ProjectId, workerHours.WorkerId));
                if (presenceHoursSum >= workerHours.NumHours)
                    workerHours.IsComplete = true;
                else
                    workerHours.IsComplete = false;
                string query = "UPDATE task_management.worker_hours " +
                    $"SET num_hours={workerHours.NumHours},is_complete={workerHours.IsComplete} " +
                    $"WHERE worker_hours_id={workerHours.WorkerHoursId};";

                ProjectService.UpdateProjectStatus(workerHours.ProjectId);
                return DBAccess.RunNonQuery(query) == 1;

            }
            catch (Exception ex)
            {
                throw ex;
            }
        }

        public static bool HasUncomletedHours(int workerId, List<int> projectIdList)
        {
            string query;
            query =
                $"SELECT COUNT(*) FROM task_management.worker_hours WHERE worker_id ={workerId} " +
                "AND is_complete=0";
            if (projectIdList != null)
            {
                query += " AND project_id IN(";
                projectIdList.ForEach(projectId =>
                {
                    query += $"{projectId},";
                });
                query = query.Substring(0, query.Length - 1);
                query += ");";
            }
            else
                query += ";";

            int count = Convert.ToInt32(DBAccess.RunScalar(query));
            return count > 0;
        }

        public static List<WorkerHours> GetAllWorkerHoursPerProject(int projectId)
        {
            string query = $"{GetWorkerHoursQuery()} WHERE w.project_id ={projectId};";
            List<WorkerHours> workerHoursList = GetWorkersHours(query);
            return workerHoursList;
        }

        public static List<WorkerHours> GetUncompletedWorkersHours(int projectId)
        {
            string query = $"{GetWorkerHoursQuery()} WHERE w.is_complete=false AND w.project_id ={projectId};";
            List<WorkerHours> workerHoursList = GetWorkersHours(query);
            return workerHoursList;
        }

        public static List<WorkerHours> GetWorkerHoursPerProject(int workerId, int projectId)
        {
            string query;
            query = $"{GetWorkerHoursQuery()} WHERE worker_id ={workerId} AND w.project_id={projectId};";
            List<WorkerHours> workerHoursList = GetWorkersHours(query);
            return workerHoursList;
        }

        public static bool AddWorkerHours(WorkerHours workerHours)
        {
            try
            {
                string query = "INSERT INTO task_management.worker_hours(project_id,worker_id,num_hours,is_complete) " +
                    $"VALUES ({workerHours.ProjectId},{workerHours.WorkerId},{workerHours.NumHours},1);";
                return DBAccess.RunNonQuery(query) == 1;
            }
            catch (Exception ex)
            {
                throw ex;
            }
        }

        public static void AddWorkerHoursToTeamProjects(User user)
        {
            if (user.TeamLeaderId != null)
            {
                List<Project> projects = ProjectService.GetProjectsInWorkingByTeamLeaderId((int)user.TeamLeaderId);
                projects.ForEach(project =>
                {
                    List<WorkerHours> workerHoursList = GetWorkerHoursPerProject(user.UserId, project.ProjectId);
                    if (workerHoursList.Count == 0)
                    {
                        WorkerHours workerHours = new WorkerHours { ProjectId = project.ProjectId, WorkerId = user.UserId };
                        AddWorkerHours(workerHours);
                    }
                });
            }
        }

        private static string GetWorkerHoursQuery()
        {
            string query = $"SELECT w.*,project_name,user_name,email,department_name " +
                $"FROM task_management.worker_hours w " +
                $"JOIN task_management.project p ON w.project_id=p.project_id " +
                $"JOIN task_management.user u ON w.worker_id=u.user_id " +
                $"JOIN task_management.department d ON u.department_id=d.department_id";
            return query;
        }

        private static List<WorkerHours> GetWorkersHours(string query)
        {
            try
            {
                Func<MySqlDataReader, List<WorkerHours>> InitWorkerHoursList = (reader) =>
                {
                    List<WorkerHours> workersHours = new List<WorkerHours>();
                    while (reader.Read())
                    {
                        WorkerHours workerHours = BaseService.InitWorkerHours(reader);
                        workersHours.Add(workerHours);
                    }
                    return workersHours;
                };

                List<WorkerHours> workerHoursList = DBAccess.RunReader(query, InitWorkerHoursList);

                return workerHoursList;
            }
            catch (Exception ex)
            {
                throw ex;
            }
        }

    }
}

File: back-end/php/task_management/bll/worker_hours_service.php
<?php

class worker_hours_service extends base_service {

    public function get_all_worker_hours($worker_id) {
        $query = $this->get_worker_hours_query() .
                " WHERE worker_id =$worker_id AND w.is_complete=0;";
        return $this->get_workers_hours($query);
    }

    public function edit_worker_hours($worker_hours) {
        $presence_hours_service = presence_hours_service::get_instance();
        $presence_hours_sum = $presence_hours_service->get_presence_hours_sum($worker_hours['projectId'], $worker_hours['workerId']);
        if ($presence_hours_sum > $worker_hours['numHours']) {
            $worker_hours['isComplete'] = true;
        } else {
            $worker_hours['isComplete'] = false;
        }
        $project_service = project_service::get_instance();
        $project_service->update_project_status($worker_hours['projectId']);
        $query = "UPDATE task_management.worker_hours " .
                "SET num_hours={$worker_hours['numHours']},is_complete={$worker_hours['isComplete']} " .
                "WHERE worker_hours_id={$worker_hours['workerHoursId']};";
        $edited = db_access::run_non_query($query) == 1;
        return $edited;
    }

    public function has_uncomleted_hours($worker_id, $project_id_list) {
        $query = "SELECT COUNT(*) as count FROM task_management.worker_hours " .
                "WHERE worker_id =$worker_id AND is_complete=0";
        if (isset($project_id_list)) {
            $query .= " AND project_id IN(";

            foreach ($project_id_list as $project_id) {
                $query .= "$project_id,";
            }
            $query = substr($query, 0, - 1);
            $query .= ");";
        } else {
            $query .= ";";
        }
        $count = db_access::run_scalar($query);
        return $count > 0;
    }

    public function get_all_worker_hours_per_project($project_id) {
        $query = $this->get_worker_hours_query() . " WHERE w.project_id =$project_id;";
        return $this->get_workers_hours($query);
    }

    public function get_uncompleted_workers_hours($project_id) {
        $query = $this->get_worker_hours_query() . " WHERE w.is_complete=false AND w.project_id =$project_id;";
        return $this->get_workers_hours($query);
    }

    public function get_worker_hours_per_project($worker_id, $project_id) {
        $query = $this->get_worker_hours_query() . " WHERE worker_id =$worker_id AND w.project_id=$project_id;";
        return $this->get_workers_hours($query);
    }

    public function add_worker_hours($worker_hours) {
        $query = "INSERT INTO task_management.worker_hours(project_id,worker_id,num_hours,is_complete) " .
                "VALUES ({$worker_hours['projectId']},{$worker_hours['workerId']},{$worker_hours['numHours']},1);";
        $created = db_access::run_non_query($query) != null;
        return $created;
    }

    public function add_worker_hours_to_team_projects($user) {
        if (isset($user['teamLeaderId'])) {
            $project_service = project_service::get_instance();
            $projects = $project_service->get_projects_in_working_by_team_leader_id($user['teamLeaderId']);
            foreach ($projects as $project) {
                $worker_hours_service = worker_hours_service::get_instance();
                $workers_hours = $worker_hours_service->get_worker_hours_per_project($user['userId'], $project['projectId']);
                if (count($workers_hours) == 0) {
                    $worker_hours = array();
                    $worker_hours['projectId'] = $project['projectId'];
                    $worker_hours['workerId'] = $user['userId'];
                    $worker_hours['numHours'] = 0;
                    $this->add_worker_hours($worker_hours);
                }
            }
        }
    }

    private function get_worker_hours_query() {
        $query = "SELECT w.*,project_name,user_name,email,department_name " .
                "FROM task_management.worker_hours w " .
                "JOIN task_management.project p ON w.project_id=p.project_id " .
                "JOIN task_management.user u ON w.worker_id=u.user_id " .
                "JOIN task_management.department d ON u.department_id=d.department_id";
        return $query;
    }

    private function get_workers_hours($query) {
        $worker_hours_list = db_access::run_reader($query, function ($model) {
                    return $this->init_worker_hours($model);
                });
        return $worker_hours_list;
    }

}
